crates/rune/src/runtime/vm.rs

```diff
@@ -977,23 +976,322 @@ impl Vm {
                     }
                 }
 
-                return Err(VmError::from(VmErrorKind::MissingField {
+                Err(VmError::from(VmErrorKind::MissingField {
                     field: field.as_str().to_owned(),
                     target: variant.type_info(),
-                }));
+                }))
             }
-            target => {
-                let hash = field.hash();
-
-                match self.call_field_fn(Protocol::SET, target, hash, (value,))? {
+            target => Ok(
+                match self.call_field_fn(Protocol::SET, target, field.hash(), (value,))? {
                     CallResult::Ok(()) => {
                         self.stack.pop()?;
                         CallResult::Ok(())
                     }
-                    result => result,
+                    CallResult::Unsupported(target) => CallResult::Unsupported(target),
+                },
+            ),
+        }
+    }
+
+    /// Implementation of getting a string index on an object-like type.
+    fn try_object_slot_index_dynamic_get(
+        &mut self,
+        target: Value,
+        string_slot: usize,
+    ) -> Result<CallResult<Value>, VmError> {
+        use crate::compile::DynamicFieldMode;
+
+        let index = self.unit.lookup_string(string_slot)?;
+        match target {
+            Value::Object(object) => {
+                let object = object.borrow_ref()?;
+
+                if let Some(value) = object.get(&***index) {
+                    return Ok(CallResult::Ok(value.clone()));
                 }
             }
-        })
+            Value::Struct(typed_object) => {
+                let typed_object = typed_object.borrow_ref()?;
+
+                if let Some(value) = typed_object.get(&***index) {
+                    return Ok(CallResult::Ok(value.clone()));
+                }
+            }
+            Value::Variant(variant) => {
+                let variant = variant.borrow_ref()?;
+
+                if let VariantData::Struct(data) = variant.data() {
+                    if let Some(value) = data.get(&***index) {
+                        return Ok(CallResult::Ok(value.clone()));
+                    }
+                }
+            }
+            target => {
+                return Ok(match target.field_mode()? {
+                    DynamicFieldMode::Never => {
+                        match self.call_field_fn(Protocol::GET, target, index.hash(), ())? {
+                            CallResult::Ok(()) => return Ok(CallResult::Ok(self.stack.pop()?)),
+                            CallResult::Unsupported(target) => CallResult::Unsupported(target),
+                        }
+                    }
+                    #[cfg(feature = "dynamic_fields")]
```

Comment:
Not a huge fan of these big featured off blocks but not much to be done about them either. Maybe consider if these sections can be cleaned up a bit (somehow) in the future. But I'm not sure how right now.